doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -14,223 +14,110 @@
+memory blocks are deallocated. This is useful when the goal is to
 
-   b3 -> b4
-   b4 -> b5
-   b5 -> b6
-   b6 -> b7
+*  amortize allocation costs for many small allocations.
+*  create better memory locality for containers.
+*  deallocate memory in bulk.
 
-   generation -u- b3
-   current -u- b3
+:class:`MemArena` has a mechanism for increasing locality by allowing coalescing the arena contents in to a single internal block of memory.
 
-A caller can actually :func:`MemArena::alloc` **any** number of bytes. Internally, if the arena is unable to allocate enough memory for the allocation, it will create a new internal block of memory large enough and allocate from that. So if the arena is allocated like:
-
-.. code-block:: cpp
-   
-   ts::MemArena *arena = new ts::MemArena(64);
-
-The caller can actually allocate more than 64 bytes. 
-
-.. code-block:: cpp
-
-   ts::MemSpan span1 = arena->alloc(16);
-   ts::MemSpan span1 = arena->alloc(256);
-
-Now, span1 and span2 are in the same generation and can both be safely used. After:
-
-.. code-block:: cpp
-
-   arena->freeze();
-   ts::MemSpan span3 = arena->alloc(512);
-   arena->thaw();
-
-span3 can still be used but span1 and span2 have been deallocated and usage is undefined. 
+Description
++++++++++++
 
-Internal blocks are adjusted for optimization. Each :class:`MemArena::Block` is just a header for the underlying memory it manages. The header and memory are allocated together for locality such that each :class:`MemArena::Block` is immediately followed with the memory it manages. If a :class:`MemArena::Block` is larger than a page (defaulted at 4KB), it is aligned to a power of two. The actual memory that a :class:`MemArena::Block` can allocate out is slightly smaller. This is because a portion of the allocated memory is reserved for the header. Another 16 bytes is reserved to track the allocation headers used by malloc; for page alignment. ie, the default block size is 32768 bytes, but it will only be able to allocate out 32720 bytes. 
+When a :class:`MemArena` is constructed it can be constructed with no allocated memory or an initial memory block of a
+specific size. The :func:`MemArena::reserve` can specific the size of the next internally allocated block if internally
+allocation should be delayed until :func:`MemArena::alloc` is called.
+
+To support coalescence and compaction of memory for better locality, the methods
+:func:`MemArena::freeze` and :func:`MemArena::thaw` are provided. Calling :func:`MemArena::freeze`
+locks down all current blocks. After this call any further allocations will be in new internal
+memory blocks. The first internal block will be large enough to hold all of the currently allocated
+memory, unless a value is passed to :func:`MemArena::freeze` to set the size explicitly. This
```

Comment:
* What I completely missed on the first reading is that `freeze` is in effect doing what `reserve` does in addition to locking down the internal blocks. It's really a freeze + reserve operation that happens.

  Suggest something like 
  ```
  Calling :func:`MemArena::freeze` locks down all current blocks and reserves
  memory for the next allocation. By default, :func:`MemArena::freeze` reserves
  a block that is large enough to hold all of the allocated memory that has
  been frozen.  If a value is passed to :func:`MemArena::freeze`, then it
  reserves a block of the given size.
  ```

* This may be outside the scope of this PR, but I would suggest removing the reservation functionality from this method, so that the pattern is
    * freeze
    * reserve
    * alloc
    * \<copy>
    * thaw

  rather than
    * freeze
    * alloc
    * \<copy>
    * thaw
  
  for clarity.